### PR TITLE
Feat/stx transfer

### LINKED ISSUE #58: Handle STX deposits in HC sample contract and event handlers
Currently, the [example hyperchain contract](https://github.com/hirosystems/stacks-hyperchains/blob/master/core-contracts/contracts/hyperchains.clar) only supports NFT and FT deposits. However, the initial testnet launch of hyperchains should also support STX deposits (particularly because hyperchain tx fees will be paid in STX in the prototype).

This also means that the deposit handlers (see #37) will need to handle STX deposits. This should be more straight-forward than FT/NFT deposits, because it is functionally just a STX mint.

### PR DESCRIPTION
**Pull request by pavitthrap**
### Description
This PR adds the ability to deposit STX into the hyperchain. 

### Applicable issues
- fixes #58

### Additional info (benefits, drawbacks, caveats)
N/A

### Checklist
- Added integration test

### CODE REVIEW

File: testnet/stacks-node/src/tests/l1_observer_test.rs [hunk 1]
```diff
@@ -539,3 +537,162 @@ fn l1_integration_test() {
...
+
+    // Start the L2 run loop.
+    let mut config = super::new_test_conf();
+    config.node.mining_key = Some(MOCKNET_PRIVATE_KEY_2.clone());
+    let miner_account = to_addr(&MOCKNET_PRIVATE_KEY_2);
+    let user_addr = to_addr(&MOCKNET_PRIVATE_KEY_1);
+    config.add_initial_balance(user_addr.to_string(), 10000000);
+    config.add_initial_balance(miner_account.to_string(), 10000000);
+
+    config.burnchain.chain = "stacks_layer_1".to_string();
```
Comment (gregorycoppola) [inline testnet/stacks-node/src/tests/l1_observer_test.rs:569]: maybe we can start grouping this under "new_test_conf"... or we can do all of them in a single follow up PR

File: testnet/stacks-node/src/tests/l1_observer_test.rs [hunk 2]
```diff
@@ -539,3 +537,162 @@ fn l1_integration_test() {
...
+
+    // Start Stacks L1.
+    let l1_toml_file = "../../contrib/conf/stacks-l1-mocknet.toml";
+    let l1_rpc_origin = "http://127.0.0.1:20443";
+
+    let mut stacks_l1_controller = StacksL1Controller::new(l1_toml_file.to_string(), true);
+    let _stacks_res = stacks_l1_controller
+        .start_process()
+        .expect("stacks l1 controller didn't start");
+    let mut l1_nonce = 0;
```
Comment (gregorycoppola) [inline testnet/stacks-node/src/tests/l1_observer_test.rs:559]: might be easier to read if you just hard-code each nonce.. but i don't feel that strongly

File: testnet/stacks-node/src/tests/l1_observer_test.rs [hunk 3]
```diff
@@ -539,3 +537,162 @@ fn l1_integration_test() {
...
+    );
+    l1_nonce += 1;
+
+    // deposit stx into hyperchains contract on L1
+    submit_tx(&l1_rpc_origin, &l1_deposit_stx_tx);
+
+    // Sleep to give the run loop time to mine a block
+    thread::sleep(Duration::from_secs(25));
+
+    // Check that the user owns STX on the hyperchain now
```
Comment (gregorycoppola) [inline testnet/stacks-node/src/tests/l1_observer_test.rs:692]: that's pretty exciting.. the user has stx on the hyperchain now?

File: src/burnchains/test.rs [hunk 4]
```diff
@@ -1299,6 +1326,96 @@ fn create_stacks_event_block_for_block_commit() {
...
+            },
+        ],
+    };
+
+    let stacks_event_block = StacksHyperBlock::from_new_block_event(&watched_contract, input);
+
+    assert_eq!(stacks_event_block.block_height, 1);
+    assert_eq!(stacks_event_block.current_block, StacksBlockId([1; 32]));
+    assert_eq!(stacks_event_block.parent_block, StacksBlockId([0; 32]));
+    assert_eq!(
```
Comment (gregorycoppola) [inline src/burnchains/test.rs:1412]: can you check the actual event that is emitted instead of just the length?

**Review by gregorycoppola — COMMENTED** [review]
thanks, Pavi... some minor comments

File: testnet/stacks-node/src/tests/l1_observer_test.rs (hunk 3 above)
Comment (pavitthrap) [reply testnet/stacks-node/src/tests/l1_observer_test.rs:692]: yup!

File: testnet/stacks-node/src/tests/l1_observer_test.rs (hunk 2 above)
Comment (pavitthrap) [reply testnet/stacks-node/src/tests/l1_observer_test.rs:559]: I personally pref this approach? We can get a third opinion here

File: testnet/stacks-node/src/tests/l1_observer_test.rs (hunk 1 above)
Comment (pavitthrap) [reply testnet/stacks-node/src/tests/l1_observer_test.rs:569]: ya good idea - I'll do that in a follow up PR

File: src/burnchains/test.rs (hunk 4 above)
Comment (pavitthrap) [reply src/burnchains/test.rs:1412]: sure, I can make this change

File: src/chainstate/stacks/db/blocks.rs [hunk 5]
```diff
@@ -4427,6 +4427,58 @@ impl StacksChainState {
...
+                                STXMintEventData {
+                                    recipient: sender,
+                                    amount,
+                                },
+                            ))
+                        });
+                        let mut execution_cost = clarity_tx.cost_so_far();
+                        execution_cost
+                            .sub(&cost_so_far)
+                            .expect("BUG: cost declined between executions");
```
Comment (kantai) [inline src/chainstate/stacks/db/blocks.rs:4462]: Same as the comment in #61 -- I think you can drop this calculation.

File: testnet/stacks-node/src/tests/l1_observer_test.rs (hunk 2 above)
Comment (kantai) [reply testnet/stacks-node/src/tests/l1_observer_test.rs:559]: Yep, I agree with the current approach -- just bumping the nonce between operations seems to make more sense.

**Review by kantai — APPROVED** [review]
LGTM!

**Pull request merged**